### PR TITLE
Improve will message

### PR DESCRIPTION
**Pull request by tigercl**
If a new Network Connection to this Session is made before the Will Delay Interval has passed, the Server MUST NOT send the Will Message.

### CODE REVIEW

File: src/emqx_delayed_publish.erl [hunk 1]
```diff
@@ -41,27 +41,21 @@
 
 -spec(load() -> ok).
 load() ->
-    emqx:hook('message.publish', {?MODULE, on_message_publish, []}).
+    emqx:hook('message.publish', {?MODULE, on_message_publish, []}),
+    emqx:hook('session.resumed', {?MODULE, on_session_resumed, []}).
 
-on_message_publish(Msg = #message{id = Id, topic = <<"$delayed/", Topic/binary>>, timestamp = Ts}) ->
+on_message_publish(Msg = #message{topic = <<"$delayed/", Topic/binary>>}) ->
```
Comment (spring2maz) [inline src/emqx_delayed_publish.erl]: worth introducing a macro for `$delayed` like `$queue` and `$share`, see `emqx.hrl`

File: src/emqx_delayed_publish.erl [hunk 2]
```diff
@@ -41,27 +41,21 @@
...
-                        true  -> error(invalid_delayed_timestamp);
-                        false -> Timestamp
-                    end
-            end,
-    PubMsg = Msg#message{topic = Topic1},
-    ok = store(#delayed_message{key = {PubAt, delayed_mid(Id)}, msg = PubMsg}),
-    {stop, PubMsg};
-
+    handle_message(undefined, binary_to_integer(Delay), Msg#message{topic = Topic1});
+on_message_publish(Msg = #message{topic = <<"$will/", Topic/binary>>, headers = #{client_id := ClientId}}) ->
```
Comment (spring2maz) [inline src/emqx_delayed_publish.erl]: and for `$will` too.

File: src/emqx_delayed_publish.erl [hunk 3]
```diff
@@ -100,6 +94,26 @@ handle_call({store, DelayedMsg = #delayed_message{key = Key}}, _From, State) ->
...
 
+handle_call({session_resumed, ClientId}, _From, State = #{timer := TRef}) ->
+    case mnesia:dirty_select(?TAB, [{#delayed_message{client = ClientId, _ = '_'}, [], ['$_']}]) of
+        [] -> 
+            {reply, ok, State};
+        WillMsgs ->
+            Result = lists:keysearch(mnesia:dirty_first(?TAB), 2, WillMsgs),
+            lists:foreach(fun(#delayed_message{key = Key}) ->
+                            mnesia:dirty_delete({?TAB, Key})
+                        end, WillMsgs),
```
Comment (spring2maz) [inline src/emqx_delayed_publish.erl]: align with `fun`
and indent more for function body.

File: src/emqx_delayed_publish.erl [hunk 4]
```diff
@@ -128,6 +142,20 @@ code_change(_OldVsn, State, _Extra) ->
 %% Internal functions
 %%------------------------------------------------------------------------------
 
+handle_message(ClientId, Delay, Msg = #message{id = Id, timestamp = Ts}) ->
```
Comment (spring2maz) [inline src/emqx_delayed_publish.erl]: IMHO. a better abstraction is to move he Delay or Timestamp logic to a separate function (which can be easily eunit-tested).
But not to make a higer level abstraction as `handle_message`
i.e. the `ok = store(...),` call can be kept at its old place, so the return value  `{stop, Msg}` tuple do not have to be scattered into multiple functions.

File: src/emqx_delayed_publish.erl [hunk 5]
```diff
@@ -39,37 +39,20 @@
...
-    PubMsg = Msg#message{topic = Topic1},
-    ok = store(#delayed_message{key = {PubAt, delayed_mid(Id)}, msg = PubMsg}),
-    {stop, PubMsg};
-
-on_message_publish(Msg) ->
-    {ok, Msg}.
+delay_publish(Msg) ->
+    delay_publish(Msg, undefined).
+
+delay_publish(Msg = #message{id = Id, timestamp = Ts, headers = #{'Will-Delay-Interval' := Interval}}, ClientId) ->
```
Comment (terry-xiaoyu) [inline src/emqx_delayed_publish.erl]: The API can not be like this.. A better design is to keep the isolation of the components. Don't depends on the internal 'Will-Delay-Interval' for the delay interval.
e.g.
`delay_publish(Msg, ID, {secs, DelaySeconds}) -> ok.`
`delay_publish(Msg, ID, {ts, Timestamp}) -> ok.`
`cancel_publish(Msg, ID) -> ok.`

File: src/emqx_delayed_publish.erl [hunk 6]
```diff
@@ -39,37 +39,20 @@
...
-                        true  -> error(invalid_delayed_timestamp);
-                        false -> Timestamp
-                    end
-            end,
-    PubMsg = Msg#message{topic = Topic1},
-    ok = store(#delayed_message{key = {PubAt, delayed_mid(Id)}, msg = PubMsg}),
-    {stop, PubMsg};
-
-on_message_publish(Msg) ->
-    {ok, Msg}.
```
Comment (terry-xiaoyu) [inline src/emqx_delayed_publish.erl]: The $delay should be kept. It is for user to publish delayed msgs via prefixed topics. 
We also need to add another prefixed topic for canceling previous published delay messages, just like the function APIs.

e.g. We may need to change the prefixed topics like this:

`$delay/<Interval or timestamp>/<id>/topic`
`$cancel_delay/<id>`

File: src/emqx_delayed_publish.erl [hunk 7]
```diff
@@ -100,6 +83,26 @@ handle_call({store, DelayedMsg = #delayed_message{key = Key}}, _From, State) ->
     ok = mnesia:dirty_write(?TAB, DelayedMsg),
     {reply, ok, ensure_publish_timer(Key, State)};
 
+handle_call({cancel_publish, ClientId}, _From, State = #{timer := TRef}) ->
+    case mnesia:dirty_select(?TAB, [{#delayed_message{client_id = ClientId, _ = '_'}, [], ['$_']}]) of
+        [] -> 
+            {reply, ok, State};
+        WillMsgs ->
+            Result = lists:keysearch(mnesia:dirty_first(?TAB), 2, WillMsgs),
```
Comment (terry-xiaoyu) [inline src/emqx_delayed_publish.erl:111]: Change the mnesia type from `set` to `ordered_set`, or this `first .. next` iteration on a set ets table will not work as expected, as set tables are not sorted in the order of insertion.

File: src/emqx_delayed_publish.erl [hunk 8]
```diff
@@ -39,37 +39,20 @@
...
-on_message_publish(Msg) ->
-    {ok, Msg}.
+delay_publish(Msg) ->
+    delay_publish(Msg, undefined).
+
+delay_publish(Msg = #message{id = Id, timestamp = Ts, headers = #{'Will-Delay-Interval' := Interval}}, ClientId) ->
+    PubAt = pub_at(Ts, Interval),
+    store(#delayed_message{key = {PubAt, delayed_mid(Id)}, client_id = ClientId, msg = Msg}).
+
+cancel_publish(ClientId) ->
```
Comment (terry-xiaoyu) [inline src/emqx_delayed_publish.erl]: Not `ClientID`, should be `ID`. User specifies a message id for its delayed message when calling this API.

**Review by terry-xiaoyu — CHANGES_REQUESTED** [review]

**Pull request closed without merging**